docs/datagen/client/modelproviders.md

```diff
@@ -1,16 +1,394 @@
-Model Providers
-===============
-The model providers are a specific type of data generators for defining models. All model providers are a subclass of `ModelProvider`.
+Model Generation
+================
+
+[Models] can be generated for models or block states by default. Each provides a method of generating the necessary JSONs (`ModelBuilder#toJson` for models and `IGeneratedBlockState#toJson` for block states). After implementation, the [associated providers][provider] must be [added][datagen] to the `DataGenerator`.
+
+Model Files
+-----------
+
+A `ModelFile` acts as the base for all models referenced or generated by a provider. Each model file stores the location relative to the `models` subdirectory and can assert whether the file exists.
+
+### Existing Model Files
+
+`ExistingModelFile` is a subclass of `ModelFile` which checks via [`ExistingFileHelper#exists`][efh] whether the model already exists within the `models` subdirectory. All non-generated models are usually referenced through `ExistingModelFile`s.
 
-`ModelProvider` provides methods to define models for blocks and items alike: cubes, single textures, doors, slabs, and even custom non-data-generated models as parent models.
+### Unchecked Model Files
 
-Implementation
+`UncheckedModelFile` is a subclass of `ModelFile` which assumes the specified model exists in some location.
+
+!!! note
+    There should be no cases where an `UncheckedModelFile` is used to reference a model. If there is, then the associated resources are not properly being tracked by `ExistingFileHelper`.
```

Comment:
Should we even mention the existence of `UncheckedModelFile` in the docs?

Reply:
I don't see a reason not to. It's the same reason why we mention `OnlyIn`: explain what it is and that it shouldn't be used.